### ISSUE TITLE
Lacking functionality enabled in native project files

### ISSUE THREAD

**Issue by PanayotCankov** [NativeScript/NativeScript#3157]
The Story
=========

Extending an app, beyond beatiful UI, requires enabling native capabilities. For iOS more often than not this happens in the xcodeproj, xcworkspace, info.plist, entitlements, storyboards etc. For Android such modifications go to the AndroidManifest.xml, build.gradle, app.gradle etc.

Status quo
==========

The native projects with the above resources reside at `platform/ios` or `platform/android`. We recomend keeping them out of source control. Our CLI is maintaining a set of build infrastructure that generates native projects out of artefacts in `App_Resources`. Not every part of a native project can be altered from `App_Resources`.

The problems
============

 - **The CLI aggregates a build system that produces native project from the App_Resources**  
   This raises the learning curve on getting started with {N}, since we can hardly relate to existing native documentation but rather must maintain our own set of articles. It increases the support cost.

 - **Incomplete implementation**  
   This filters some of the functionality that native project can easily achieve. This results in 
    - Apps keeping project files from the `platform` folder in source contol countrary to the recommended approach
    - Writing custom hooks to setup the native projects
   Please note that App_Resources maintain two different sets of iOS and Android artefacts so it does not provide in any way a cross-platform abstraction.

 - **We can not use the native tools to edit these project files**  
   Sometimes the modifications required to enable push notifications can be achieved easily using the UI of the native tools (Xcode and AndroidStudio) but since changes at `platforms` are not propagated back to the App_Resources this is a no-go. Also changes done in the native project files may be overwriten by the CLI with the settings in App_Resources.

 - **The community tries to handle lacking functionality**  
   In general it is good to have a live and vivid community but there is a risk that multiple plugins that try to implement the lacking functionality may become incompatible together. As an example the nativescript-plugin-firebase tries to handle the lack of .entitlements support in App_Resources by copying and setting the appopriate Xcode build flags using hooks, but may as well overwrite similar setup from other plugins. While this plugin is in development and there is no doubt it will evolve in a robust solution, it may be better to handle this in the core teams.

 - **Clients had problems with lacking functionality at App_Resources**
   Clients had the same class of problems along the lifetime of {N} when trying to automate builds or share the app between multiple developers when using some features that are not yet implemented through the App_Resources.
    - Problems adding icons and launch screens in xcassets generated by third party tools
    - Problems setting up api keys that have to be added in the native projects
    - Setting up entitlements
    - Setting up additional build flags

One possible solution
==================

One possible solution would be to embrace a change of hearth and let the apps officially keep most of the project files from `platform/ios` and `platform/android` in source control. The `app` folders in the native projects should be still ignored in source control and their content will still be generated upon build by the CLI from the JavaScript files. Intermediate project files and temporary output should still be ignored. This will probably have to ship with carefully written .gitignore.

Pros:
 - Xcode and AndroidStudio can be used to enable native capabilities in the native projects
 - Source control roundtrip and CI may be easier
 - Removing the build steps that move files from App_Resources to platform/*
 - Removing the required documentation on how App_Resources works
 - Apps and Plugins will be allowed to more directly reuse existing articles/skills with native projects
 - It may be easier in future to add {N} to existing native projects

Cons:
 - If the integration points for the runtimes in a native project are not stable enough it would be hard to upgrade existing projects.  
   This will require the runtimes to elaborate on how are their bits added to a native app, and if this is feasible.

   (Perhaps I am biased at the moment I need more input for the Cons part)

Non complete task list:
  - [ ] This may invalidate the bundle id that is stored in the package.json and the CLI may need updates to read package and bundle ids from the native project files (Info.plst and AndroidManifest.xml in platforms) instead of pushing them from the package.json to these files on build.
 - [ ] We will have to revisit the project scafolding in the CLI and the default temaplte.
- [ ] The Android runtime project and tools have to be aligned somewhat with the ones produced by AndroidStudio.
- [ ] For plugins that still need to apply changes to the native projects we will have to provide guidance
- [ ] We will have to reorganize some of the existing documentation

**Comment by rosen-vladimirov** [comment]
Great explanation of the problem, however I would like to suggest another approach. 

Instead of using Android Studio and XCode, why don't we find a way to extend the projects, for example via CLI hooks or some other way. We should check how Cordova resolves the same issue - the main power of Cordova are plugins, which allow you to extend the functionality of your project. In our case it looks like the plugins are not able to change the native project in a way they need it, but does it mean that I need Android Studio in order to build my project?

The `App_Resources` issue - currently you have the images in two separate directories in your application. I'm not Android developer, but for me it's pretty easy to change the default images there. For iOS it's the same. Your suggestion requires to point our users to read XCode and Android Studio documentation how to generate images. So if I'm a web developer, I'll have to learn how to use NativeScript, how to use Android Studio, how to use XCode... Where's the cross-platform solution here? Why do I need any other tool for working with NativeScript project, shouldn't NativeScript CLI be enough? If working with App_Resources is difficult, why don't we introduce commands that will help in such cases, why should we go to native projects at all?

One more note: asking users to set project id in Info.plist, AndroidManifest.xml, app.gradle or whatever non-JS typical file sounds like skipping the cross-platform solution for me, but maybe it's just me.

So to summarize, I'm strongly against the suggested solution. As the mentioned problems are valid, my suggestion is to:
 - provide better extensibility model for projects, similar to Cordova's behavior, where plugins can extend pretty much everything. 
 - Check what's missing and difficult to use with App_Resources and improve documentation and/or add NativeScript CLI commands that will help resolving the problems.

**Comment by PanayotCankov** [comment]
I am not sure I am convinced we need CLI commands and scripts for everything. It is a lot of code to maintain in a single place.

On some of your questions:

 - You will not be forced in any way to use Android Studio nor Xcode (emphasis on the lower case C). You will simply not be denied the possibility to use Android Studio or Xcode if you want to.
 - We are going to sink Cordova because we are better, not because we are going to copy them.
 - Plugins currently change files in the platform/ios and platform/android upon prepare but the developers have no mechanism to control what changes should and what changes should not be applied to their native project. The plugins will still be recommended to plug-and-play by modifying the files in platform/ios and platform/android upon install. But since these files will be part of source control this will greatly increase the visibility of what a plugin does to your native project.
 - Third party tools (icon generators come to mind first) will still work with the projects.
 - "Where's the cross-platform solution here?". We still don't have cross platform solution here. And when we does a tool that handles icons for {N} will simply be a dev dependency that install its own npm command, given a path to 1024x1024 icon, it will update the AndroidManifest.xml, the iOS Info.plist and put the images in the proper places in the native project.
 - The CLI is rarely enough. You use at least vim and git with it.
 - Command such as `tns set id com.telerik.demoapp` can easily set the id in both the AndroidManifest.xml and Info.plist for the user, keeping it cross platform. This is in case it is easier to educate the user that command. We can argue which is easier, finding out that in native project the Info.plist and AndroidManifest.xml keep the id or that the "id" in the package.json of a native script project will patch your Info.plsit AndroidManifest.xml from App_Resources on their way to platform/android and platform/ios.
 - This will not harm the ability to automate the plugin installation in anyway. Plugins should be able to extend pretty much everything. If we keep an ordinary native project in source control plugins will need an XML reader/writer, probably the `xcode` package and a script or two in their package.json. And probably some imagination. Lowering the learning curve by eliminating the $injector, services and hooks part of the CLI.
 - Trying to address everything that is missing in App_Resources will eventually lead to a solution that can keep everything a native project can in App_Resources. But it won't be a native project. We will have to educate the developers what the differences are. And so far we are constantly a step behind from what native project can offer and what our customers need.

**Comment by KristinaKoeva** [comment]
This seems related https://github.com/NativeScript/nativescript-ios-imessages/issues/3

**Comment by PanayotCankov** [comment]
@KristinaKoeva All extensions are - sharing, today widgets, keyboards, watch apps, etc.

**Comment by atanasovg** [comment]
@PanayotCankov What's the state of this feature and is it still a valid problem?

**Comment by frapaa** [comment]
@PanayotCankov This is something I'm facing now. What was the conclusion of this issue?
